internal/lsp/server.go

```diff
@@ -37,24 +38,31 @@ func SendNotification(method string, params interface{}) {
 
 func RunServer[State any](args ServerArgs[State]) {
 	buf := NewMessageBuffer(os.Stdin)
+	mu := sync.Mutex{}
 
 	for {
 		request := buf.Read()
 
-		bytes, err := json.Marshal(args.Handler(request, &args.InitialState))
-		if err != nil {
-			panic(err)
-		}
-
-		rawMsg := json.RawMessage(bytes)
-		encoded := encodeMessage(jsonrpc2.Response{
-			ID:     request.ID,
-			Result: &rawMsg,
-		})
-		_, err = fmt.Print(encoded)
-		if err != nil {
-			panic(err)
-		}
+		go func() {
+			bytes, err := json.Marshal(args.Handler(request, &args.InitialState))
+			if err != nil {
+				panic(err)
+			}
+
+			rawMsg := json.RawMessage(bytes)
+			encoded := encodeMessage(jsonrpc2.Response{
+				ID:     request.ID,
+				Result: &rawMsg,
+			})
+
+			mu.Lock()
```

Comment:
Why take a lock before doing a fmt.Print?

Reply:
In order to avoid concurrent writes to stdout which would mix the outputs, thus making the response(s) unreadable
(Context: language server protocol communicates via a json-rpc API over stdio and stdout)